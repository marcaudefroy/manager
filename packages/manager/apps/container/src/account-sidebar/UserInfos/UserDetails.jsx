import React from 'react';
import PropTypes from 'prop-types'

/**
 * Displays basic information about a user: its nic, organization and email.
 * @component
 */
const UserDetails = ({ user, cssBaseClassName }) => {
  return (
    <p>
    { user.organisation && (
      <span
        className={`d-block ${cssBaseClassName}_text-small`}
      >
        { user.organisation }
      </span>
    )}
      <span
        className={`d-block ${cssBaseClassName}_text-small text-break`}
      >
          { user.email }
      </span>
    { user.email !== user.nichandle && (
      <span
        className={`d-block ${cssBaseClassName}_text-small`}
      >
        { user.nichandle }
      </span>
    )}
    </p>
  );
};

UserDetails.propTypes = {
  user: PropTypes.shape({
    email: PropTypes.string,
    nichandle: PropTypes.string,
    organisation: PropTypes.string,
  }).isRequired,
  cssBaseClassName: PropTypes.string,
};

UserDetails.defaultProps = {
  user: {},
};

export default UserDetails;
