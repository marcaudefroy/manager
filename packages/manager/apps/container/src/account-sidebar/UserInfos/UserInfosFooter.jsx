import React from 'react';
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next';
import { useShell } from '@/context';

import { TRANSLATE_NAMESPACE } from '../constants';

/**
 * Displays a footer which allows to logout.
 * @component
 */
const UserInfosFooter = ({ cssBaseClassName, translationBase }) => {
  const { t } = useTranslation(TRANSLATE_NAMESPACE);
  const shell = useShell();

  const logoutHandler = () => {
    shell.getPlugin('tracking').trackClick({
      name: 'hub::sidebar::profile::go-to-log-out',
      type: 'action',
    });
    shell.getPlugin('auth').logout();
  };

  return (
    <div className={`text-left ${cssBaseClassName}_links`}>
      <hr className="my-1" />
      <button
        aria-label={`${translationBase}_footer_logout`}
        type="button"
        role="button"
        className="btn btn-link"
        data-navi-id="logout"
        onClick={logoutHandler}
      >
        {t(`${translationBase}_footer_logout`)}
      </button>
    </div>
  );
};

UserInfosFooter.propTypes = {
  cssBaseClassName: PropTypes.string,
  translationBase: PropTypes.string,
};

export default UserInfosFooter;
