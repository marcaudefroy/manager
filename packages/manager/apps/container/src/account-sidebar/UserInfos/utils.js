import { EXCLUDED_ROLES } from './constants';

/**
 * Formats the user name as initials.
 * @param {Object} user
 * @returns {string} Formatted user initials
 */
export function getUserNameInitials(user) {
  return user.firstname && user.name
    ? `${user.firstname[0]}${user.name[0]}`.toUpperCase()
    : '';
}

/**
 * Formats the user name as a displayable name.
 * @param {Object} user
 * @returns {string} Formatted user name
 */
export function getUserDisplayName(user) {
  return `${user.firstname} ${user.name}`;
}

/**
 * Returns the support level of the user.
 * @param {Object} user
 * @returns {string} user support level
 */
export function getSupportLevel(user) {
  return user.supportLevel;
}

/**
 * Check whether or not the user is flagged as trusted.
 * @param {Object} user
 * @returns {boolean}
 */
export function isTrustedUser(user) {
  return user.isTrusted;
}

/**
 * Returns the role associated to the user.
 * @param {Object} user
 * @returns {string}
 */
export function getUserRole(user) {
  return !EXCLUDED_ROLES.includes(user?.auth?.method) ? user?.auth?.method : '';
}
