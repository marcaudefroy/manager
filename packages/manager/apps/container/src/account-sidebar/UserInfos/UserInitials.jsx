import React from 'react';
import PropTypes from 'prop-types'

import { getUserNameInitials } from './utils';

/**
 * Displays user's initials.
 * @component
 */
const UserInitials = ({ user, cssBaseClassName }) => {
  return (
    <span
      className={`${cssBaseClassName}_initials ${cssBaseClassName}_initials-centered`}
    >
      { getUserNameInitials(user) }
    </span>
  );
};

UserInitials.propTypes = {
  user: PropTypes.shape({
    firstname: PropTypes.string,
    name: PropTypes.string,
  }).isRequired,
  cssBaseClassName: PropTypes.string,
};

UserInitials.defaultProps = {
  user: {},
};

export default UserInitials;
