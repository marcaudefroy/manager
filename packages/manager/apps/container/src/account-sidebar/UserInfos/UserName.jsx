import React from 'react';
import PropTypes from 'prop-types'

import { getUserDisplayName } from './utils';

/**
 * Displays user's name.
 * @component
 */
const UserName = ({ user, cssBaseClassName }) => {
  return (
    <p className={`${cssBaseClassName}_profile_link mb-1`}>
      { getUserDisplayName(user) }
    </p>
  );
};

UserName.propTypes = {
  user: PropTypes.shape({
    firstname: PropTypes.string,
    name: PropTypes.string,
  }).isRequired,
  cssBaseClassName: PropTypes.string,
};

UserName.defaultProps = {
  user: {},
};

export default UserName;
