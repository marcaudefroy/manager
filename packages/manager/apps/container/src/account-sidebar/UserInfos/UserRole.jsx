import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next';

import { getUserRole } from './utils';
import { TRANSLATE_NAMESPACE } from '../constants';

/**
 * Displays user's role.
 * @component
 */
const UserRole = ({ user, cssBaseClassName, translationBase }) => {
  const { t } = useTranslation(TRANSLATE_NAMESPACE);
  const [ role, setRole ] = useState('');

  useEffect(() => {
    setRole(getUserRole(user));
  }, [ user ])

  return (
    <div
      className={`${cssBaseClassName}_user-name mb-2`}
      data-navi-id="account-sidebar-block"
    >
     { role && (
        <div>
          <span
              className={`${cssBaseClassName}_role oui-badge oui-badge_warning`}
          >
            { t(`${translationBase}_role_${role}`) }
          </span>
        </div>
      )}
    </div>
  );
};

UserRole.propTypes = {
  user: PropTypes.shape({
  }).isRequired,
  cssBaseClassName: PropTypes.string,
  translationBase: PropTypes.string,
};

UserRole.defaultProps = {
  user: {},
};

export default UserRole;
