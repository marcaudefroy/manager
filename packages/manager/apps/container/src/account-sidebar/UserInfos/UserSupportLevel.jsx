import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next';

import { isTrustedUser, getSupportLevel } from './utils';
import { TRANSLATE_NAMESPACE } from '../constants';

/**
 * Displays user's support level.
 * @component
 */
const UserSupportLevel = ({ user, cssBaseClassName, translationBase }) => {
  const { t } = useTranslation(TRANSLATE_NAMESPACE);
  const [ level, setLevel ] = useState(getSupportLevel(user));

  useEffect(() => {
    setLevel(getSupportLevel(user));
  }, [ user ]);

  return (
    <p className="oui-chip mb-1">
      <span className={`${cssBaseClassName}_text-small`}>
        {t(
          `${translationBase}_support_level_${level}${
            isTrustedUser(user) ? '_trusted' : ''
          }`,
        )}
      </span>
    </p>
  );
};

UserSupportLevel.propTypes = {
  user: PropTypes.shape({
    auth: PropTypes.shape({
      method: PropTypes.string,
    }),
  }).isRequired,
  cssBaseClassName: PropTypes.string,
  translationBase: PropTypes.string,
};

UserSupportLevel.defaultProps = {
  user: {},
  translationBase: 'user_infos',
};

export default UserSupportLevel;
