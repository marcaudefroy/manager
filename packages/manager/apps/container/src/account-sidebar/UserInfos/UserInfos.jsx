import React from 'react';
import { useTranslation } from 'react-i18next';
import { buildURL } from '@ovh-ux/ufrontend';

import { useShell, useUser } from '@/context';
import UserInitials from './UserInitials';
import UserSupportLevel from './UserSupportLevel';
import UserName from './UserName';
import UserRole from './UserRole';
import UserDetails from './UserDetails';
import UserInfosFooter from './UserInfosFooter';

import { TRANSLATE_NAMESPACE } from '../constants';

const UserInfos = () => {
  const { t } = useTranslation(TRANSLATE_NAMESPACE);
  const shell = useShell();
  const user = useUser();

  const cssBaseClassName = 'manager-account-sidebar-user-infos';
  const translationBase = 'user_infos';
  const userAccountURL = shell
    .getPlugin('navigation')
    .getURL('dedicated', '#/useraccount/dashboard');

  const userAccountClickHander = () =>
    shell.getPlugin('tracking').trackClick({
      name: 'hub::sidebar::profile::go-to-account',
      type: 'action',
    });

  return (
    <div className={`${cssBaseClassName} mb-3`}>
      <a
        className={`${cssBaseClassName}_profile`}
        aria-label={t(`${translationBase}_manage_my_account`)}
        href={userAccountURL}
        target="_top"
        onClick={userAccountClickHander}
      >
        <UserInitials user={user} cssBaseClassName={cssBaseClassName} />
        <UserSupportLevel
          user={user}
          cssBaseClassName={cssBaseClassName}
          translationBase={translationBase}
        />
        <UserName user={user} cssBaseClassName={cssBaseClassName} />
      </a>
      <UserRole
        user={user}
        cssBaseClassName={cssBaseClassName}
        translationBase={translationBase}
      />
      <UserDetails user={user} cssBaseClassName={cssBaseClassName} />
      <UserInfosFooter
        cssBaseClassName={cssBaseClassName}
        translationBase={translationBase}
      />
    </div>
  );
};

export default UserInfos;
