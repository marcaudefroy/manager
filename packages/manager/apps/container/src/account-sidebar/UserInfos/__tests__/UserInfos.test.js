import React from 'react';

import { renderWithShell } from '../../../__tests__';

import UserInfos from '../UserInfos';

describe('UI testing of UserInfos', () => {
  it('renders correctly', async () => {
    const user = {
      firstname: 'Tester',
      name: 'testee',
      supportLevel: 1,
    };
    const universe = 'web';

    const environment = {
      getUser: () => user,
      getUniverse: () => universe,
      getUserLocale: () => 'fr_FR',
      getRegion: () => 'EU',
    };

    const { asFragment } = await renderWithShell(<UserInfos />, {
      environment,
    });

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('UserInfos', () => {
  it('displays informations', async () => {
    const user = {
      firstname: 'Guybrush',
      name: 'Threepwood',
      supportLevel: 'standard',
      organisation: 'OVH',
      email: 'guybrush.threepwood@email.com',
      nichandle: 'gt-nichandle',
      auth: { method: 'my-role' },
    };
    const universe = 'web';

    const environment = {
      getUser: () => user,
      getUniverse: () => universe,
      getUserLocale: () => 'fr_FR',
      getRegion: () => 'EU',
    };

    const { queryByText } = await renderWithShell(<UserInfos />, {
      environment,
    });

    expect(queryByText('user_infos_support_level_standard')).toBeTruthy(); // support level
    expect(queryByText('Guybrush Threepwood')).toBeTruthy(); // name
    expect(queryByText('GT')).toBeTruthy(); // initials
    expect(queryByText('OVH')).toBeTruthy(); // organisation
    expect(queryByText('guybrush.threepwood@email.com')).toBeTruthy(); // email
    expect(queryByText('gt-nichandle')).toBeTruthy(); // nichandle
  });
});
