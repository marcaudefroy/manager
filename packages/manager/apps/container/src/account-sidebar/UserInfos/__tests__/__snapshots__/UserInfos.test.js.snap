// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI testing of UserInfos renders correctly 1`] = `
<DocumentFragment>
  <div
    class="manager-account-sidebar-user-infos mb-3"
  >
    <a
      aria-label="user_infos_manage_my_account"
      class="manager-account-sidebar-user-infos_profile"
      href="http://localhost/#/useraccount/dashboard"
      target="_top"
    >
      <span
        class="manager-account-sidebar-user-infos_initials manager-account-sidebar-user-infos_initials-centered"
      >
        TT
      </span>
      <p
        class="oui-chip mb-1"
      >
        <span
          class="manager-account-sidebar-user-infos_text-small"
        >
          user_infos_support_level_1
        </span>
      </p>
      <p
        class="manager-account-sidebar-user-infos_profile_link mb-1"
      >
        Tester testee
      </p>
    </a>
    <div
      class="manager-account-sidebar-user-infos_user-name mb-2"
      data-navi-id="account-sidebar-block"
    />
    <p>
      <span
        class="d-block manager-account-sidebar-user-infos_text-small text-break"
      />
    </p>
    <div
      class="text-left manager-account-sidebar-user-infos_links"
    >
      <hr
        class="my-1"
      />
      <button
        aria-label="user_infos_footer_logout"
        class="btn btn-link"
        data-navi-id="logout"
        role="button"
        type="button"
      >
        user_infos_footer_logout
      </button>
    </div>
  </div>
</DocumentFragment>
`;
